### ISSUE TITLE
$$$$temp$$$$ bug/error on decompilation.

### ISSUE THREAD

**Issue by Animelici804** [krzys-h/UndertaleModTool#164]
https://cdn.discordapp.com/attachments/571263675571372034/571992794323746817/unknown.png
A decompiling error/bug.
Please fix it. Kneesnap said I can make an issue in this page, so I did it.

**Comment by colinator27** [comment]
This isn't technically a decompilation bug, because that's meant to be there, but we should get rid of it. I believe it appears when you're inside of a `with` or `switch` statement and you return out of it with a value.

**Comment by colinator27** [comment]
Confirmed this. With this test code, for example:
```
with (all)
{
    if (0 + 1)
        return 1;
    x = 1;
}
```

The disassembly is:
```
.localvar 0 arguments
.localvar 1 $$$$temp$$$$ 3
00000: pushi.e -3
00001: pushenv 00013
00002: pushi.e 1
00003: conv.i.v
00004: pop.v.v local.$$$$temp$$$$
00006: popenv [drop]
00007: push.v local.$$$$temp$$$$
00009: ret.v
00010: pushi.e 1
00011: pop.v.i self.x
00013: popenv 00002
```

And the decompilation is:
```
var $$$$temp$$$$;
with(all)
{
    $$$$temp$$$$ = 1
    return $$$$temp$$$$
    x = 1
}

```

As you can see in the disassembly, it exits the `with` context and then returns the value from the local variable, so as to make sure it returns the correct value from the right context.

**Comment by colinator27** [comment]
This would be especially worthwhile if the code was more like this:
```
with (obj)
{
    if (abc)
        return x;
    y = 15;
}
```
It needs to return `obj.x` and not `self.x`, so it assigns `$$$$temp$$$$`, leaves the `obj` context, then returns. Keep in mind that local variable scopes last throughout the entire script or event, so this works I guess. I don't know why they didn't just push onto the stack or something like that, or maybe that's impossible with the `pushenv` and `popenv` instructions? Just seems like a lazy workaround to me.

**Comment by colinator27** [comment]
Actually, another thing of note. My compiler right now doesn't generate $$$$temp$$$$. It just pushes it onto the stack, runs `popenv`, and then returns. I'll check to see if that doesn't work.

**Comment by colinator27** [comment]
Okay, that does NOT work. My compiler generates something more like:
```
with(all)
{
    if 1
    {
        _temp_local_var_1 = 1
        // leave with context
        return _temp_local_var_1
    }
    x = 1
}
```

But apparently it seems that the stacks are different, or are at least cleared. That's pretty strange. I guess I need to fix that then.

**Comment by colinator27** [comment]
Also, I'm assuming that for switch statements, it will do the same thing but clean up in its own way (the equivalent of `popenv`, but probably just for cleaning up the stack). Actually, this might happen for other loops like `repeat`. Maybe we should look into that...

**Comment by Animelici804** [comment]
Okay.

**Comment by Kneesnap** [comment]
It might help if we had the script you were trying to decompile, @Animelici804.

**Comment by Animelici804** [comment]
Okay

**Comment by Animelici804** [comment]
The game maker export script?

**Comment by colinator27** [comment]
Have a fix for this locally now- it just detects
```
$$$$temp$$$$ = <>
return $$$$temp$$$$
```
and changes it into
```
return <>
```
as you would expect.

**Comment by colinator27** [comment]
Also gets rid of the local var definition in the beginning, always. So if there's any other occurrences of $$$$temp$$$$ then it'd decompile slightly incorrect, but that's unlikely.

**Comment by colinator27** [comment]
Fixed in #162

**Comment by Animelici804** [comment]
Ok then, I can close the issue now.
